### PR TITLE
[AIRFLOW-6489] Separate integrations in tests

### PR DESCRIPTION
**Pull request by potiuk**
You can now choose which integration you want to start when you run
Breeze as well as when CI tests are run. Now by default Breeze
and CI runs without integrations, but you can add them via
Breeze flags or by environment variables when CI is executed.

We have pytest markers now that mark tests that can be run for
integration, backend and runtime selected.

Also we have now more test jobs - we have separate test run
for all non-integration tests (with less memory used by the
integrations) and separate jobs that run integration tests
only (more memory used for integrations but far less number
of tests to run)

---
Issue link: [AIRFLOW-6489](https://issues.apache.org/jira/browse/AIRFLOW-6489/)

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Review by turbaszek — APPROVED** [review]

**Comment by potiuk** [comment]
Hello everyone. I am still running final tests but I think this one is ready to review. I think - regardless of our Github Actions move, it will be super-helpful to have faster, more stable tests and especially - the local dev environment with Breeze is now really, really convenient to use (far lest resource usage without extra integrations, full pytest marker integration etc.).

**Comment by potiuk** [comment]
@mik-laj @ashb @dimberman @kaxil @feluelle @Fokko @nuclearpinguin -> please take a look.

File: TESTING.rst [hunk 1]
```diff
@@ -152,46 +152,192 @@ You can also specify individual tests or a group of tests:
...
+
+   * - Integration
+     - Description
+   * - cassandra
+     - Integration required for Cassandra hooks
+   * - kerberos
+     - Integration that provides Kerberos authentication
+   * - mongo
+     - Integration required for MongoDB hooks
+   * - openldap
```
Comment (potiuk) [inline TESTING.rst:192]: Note to reviewers: I could not find any openldap integration tests. Maybe some of them are simply skipped now. I'd love if you can point me to the tests that should be marked with pytest.mark.integration("openldap")

**Comment by potiuk** [comment]
Rearranged documentation slightly.

File: TESTING.rst [hunk 2]
```diff
@@ -152,46 +152,262 @@ You can also specify individual tests or a group of tests:
...
+     - Integration required for OpenLDAP hooks
+   * - rabbitmq
+     - Integration required for Celery executor tests
+   * - redis
+     - Integration required for Celery executor tests
+
+Below command starts mongi integration only:
```
Comment (mik-laj) [inline TESTING.rst]:
```suggestion
Below command starts mongo integration only:
```

File: scripts/ci/in_container/_in_container_utils.sh [hunk 3]
```diff
@@ -195,4 +195,5 @@ EOF
 
 function stop_output_heartbeat() {
     kill "${HEARTBEAT_PID}"
+    wait "${HEARTBEAT_PID}" || true 2> /dev/null
```
Comment (mik-laj) [inline scripts/ci/in_container/_in_container_utils.sh:198]: What do you think to start using `jobs/fg/bg/kill %1` here? Running scripts this way is a bit hacky to me.

File: scripts/ci/in_container/krb5/krb5.conf [hunk 4]
```diff
@@ -17,8 +17,8 @@
 # under the License.
 
 [logging]
-default = FILE:/var/log/krb5libs.log
-kdc = FILE:/var/log/krb5kdc.log
+default = FILE:/var/log/kerberos.log
```
Comment (mik-laj) [inline scripts/ci/in_container/krb5/krb5.conf]: Why do we want to have these logs in one file?

**Review by mik-laj — APPROVED** [review]

File: tests/providers/apache/cassandra/hooks/test_cassandra_hook.py [hunk 5]
```diff
@@ -30,15 +31,7 @@
...
-    try:
-        Cluster(["cassandra"])
-        return False
-    except UnresolvableContactPoints:
-        return True
-
-
-@unittest.skipIf(cassandra_is_not_up(), "Cassandra is not up.")
+@pytest.mark.integration("cassandra")
 class TestCassandraHook(unittest.TestCase):
```
Comment (turbaszek) [inline tests/providers/apache/cassandra/hooks/test_cassandra_hook.py:35]: If the reason behind Cassandra flakiness was lack of resources then this is should work.

**Review by turbaszek — APPROVED** [review]

File: scripts/ci/in_container/_in_container_utils.sh (hunk 3 above)
Comment (potiuk) [reply scripts/ci/in_container/_in_container_utils.sh:198]: Kill %1 is the same as PID but it is really intended for interactive use (when you see the output of `jobs` command). You never know which job you are killing this way.  If you have some other background jobs started before your  "start_output_heartbeat" then `kill %1` will kill the other job, not the heartbeat. Storing PID of the started process and using it explicitly is the only sure way I know to kill exactly the process I intend to.

File: scripts/ci/in_container/krb5/krb5.conf (hunk 4 above)
Comment (potiuk) [reply scripts/ci/in_container/krb5/krb5.conf]: Mistake ! Thanks. I will fix it.

File: tests/providers/apache/cassandra/hooks/test_cassandra_hook.py (hunk 5 above)
Comment (potiuk) [reply tests/providers/apache/cassandra/hooks/test_cassandra_hook.py:35]: I truly hope so! It looks like so far.

**Review by dimberman — APPROVED** [review]
This is awesome, thank you @potiuk !

**Pull request merged**